### ISSUE TITLE
New function to allow easy paralelization over vmap()

### ISSUE THREAD

**Issue by MRiabov** [jax-ml/jax#24547]
Please:

- [CHECKED] Check for duplicate requests.
- [x] Describe your goal, and if possible provide a code snippet with a motivating example.

Hello Jax team,
I have a simple request:

I've read about all the padding and even shape_poly, but I have the following issue: I have jnp.ndarrays of shape like this: [(50, 6, 6), (100, 6, 6), (1000, 6, 6), (1,6,6)]. Array shape will be known at compile time and will remain consistently the same, but each individual array will have different ndarray. 
For each array, the function is already vmapped. I understand I could:
```
vmapped_func = vmap(func_to_compute)
scan(vmapped_func, (), [(50, 6, 6), (100, 6, 6), (1000, 6, 6), (1,6,6)])
```
Which works, but it requires sequential execution.

Jax could have a simple function to parallelize inputs of lists, kind of like #paralel_for in C++ MPI.

Why not? It's something that would speed up compute for "jagged" arrays.

P.S. I see now that I could actually concatenate and then slice the arrays to enable paralel execution, but nevertheless - why do it manually?

**Comment by lockwo** [comment]
What's the disadvantage of concatenation? In general, jax doesn't support much in terms of operating things like vmap over non array datatypes (generally prioritizing SoA over AoS).

**Comment by MRiabov** [comment]
> What's the disadvantage of concatenation? In general, jax doesn't support much in terms of operating things like vmap over non array datatypes (generally prioritizing SoA over AoS).

updated the last words to "Why do it manually"?

Because... why do it manually? it's a util function everybody can use, so that, in the future, hundreds or thousands of devs like me don't have to google [jax ragged array](https://www.google.com/search?q=jax+ragged+array&oq=jax+ragge&gs_lcrp=EgZjaHJvbWUqDggAEEUYJxg7GIAEGIoFMg4IABBFGCcYOxiABBiKBTIGCAEQRRg5MgcIAhAAGIAEMggIAxAAGBYYHjINCAQQABiGAxiABBiKBTIGCAUQRRg8MgYIBhBFGDwyBggHEC4YQNIBCTEwNzk2ajBqMagCALACAA&sourceid=chrome&ie=UTF-8), and inevitably hitting a wall.

Basically...
```
positions=cumsum(map(lambda array: array.shape[0]))
concatted_array=jnp.concat(array for array in arrays])
results=vmap(func)(concatted_array)
slice(results, positions) # I don't know which func does that.
```

This would literally save hours for anybody who is looking how to solve this problem. I was looking for this for 3(!) months, and only now an coming with solution,

**Comment by lockwo** [comment]
Adding one function to consume python lists (or sequences), would be a major break from the design of the API so far. Especially considering concatenating arrays is a pretty core subroutine that is standard practice for working with vmap. It seems like such utility functions might belong better in a third party repo (e.g. in one of the deepmind ecosystem like chex, or equinox, or one of the many pytree utility packages) rather than in jax core.

**Comment by jakevdp** [comment]
(xref: https://stackoverflow.com/questions/79123001/storing-and-jax-vmap-over-pytrees)

I think we probably won't add anything like the proposed API. A core philosophy of JAX is to keep the API surface small, and achieve functionality through composition. Here, as you showed, you can achieve the desired functionality in a few lines by composing `concatenate` and `vmap`. That seems like a very good solution to the problem – what do you think?

**Comment by jakevdp** [comment]
I'm going to close this, since it's been inactive for a while. Feel free to chime back in here if you have other questions!
